### PR TITLE
Provide function to pivot label DataFrame for gwas functions

### PR DESCRIPTION
**Pull request by henrydavidge**
Signed-off-by: Henry D <henrydavidge@gmail.com>

## What changes are proposed in this pull request?

It's a little bit tricky to pivot labels (is this a better name than phenotypes in general?) into the format that Glow expects for is gwas functions. This PR provides a function to do this pivoting for both the loco and non-loco cases. I'll open a separate PR that updates the docs and our gwas notebooks.

## How is this patch tested?
- [x] Unit tests
- [ ] Integration tests
- [ ] Manual tests

(Details)

### CODE REVIEW

File: python/glow/wgr/tests/test_pivot_for_gwas.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        0.2696910984588018, -0.5467797324110817, 1.6258456800754162, 0.03999127669745301,
+        1.093584890813106, -0.5203782912574525, -0.6729752808261074, -0.3071464501504048,
+        -0.016041461027687984, 0.3213354383996257
+    ]
+    assert np.allclose(expected, pt_list[:10])
+
+
+def test_pivot_no_loco(spark):
+    df = pd.read_csv('test-data/wgr/ridge-regression/pts.csv')
+    df = df.set_index('sample_id')
```
Comment (karenfeng) [inline python/glow/wgr/tests/test_pivot_for_gwas.py:36]: Instead of re-setting `df` here, can you `set_index` immediately following the `read_csv` or use `set_index(inplace=True)`?

File: python/glow/wgr/tests/test_pivot_for_gwas.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+def test_error_too_many_levels(spark):
+    df = pd.DataFrame(columns=['c1', 'c2', 'c3'])
+    df = df.set_index(['c1', 'c2', 'c3'])
+    with pytest.raises(ValueError):
+        functions.pivot_for_gwas(spark, df)
+
+
+def test_pivot_loco(spark):
+    df = pd.read_csv('test-data/wgr/ridge-regression/level1YHatLoco.csv')
+    df = df.set_index(['sample_id', 'contigName'])
```
Comment (karenfeng) [inline python/glow/wgr/tests/test_pivot_for_gwas.py:20]: Instead of re-setting `df` here, can you `set_index` immediately following the `read_csv` or use `set_index(inplace=True)`? Also, we likely expect the sample IDs and contig names to be strings, which can be done as follows:
```
level1_yhat_loco_df = pd.read_csv(f'{data_root}/level1YHatLoco.csv',
                                  dtype={
                                      'sample_id': 'str',
                                      'contigName': 'str'
                                  }).set_index(['sample_id', 'contigName'])
```
We may want to add a check for this in the pivoting function.

File: python/glow/wgr/tests/test_pivot_for_gwas.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    pivoted = functions.pivot_for_gwas(spark, df)
+    assert pivoted.count() == 12
+    assert pivoted.columns == ['label', 'contigName', 'values']
+    pt_list = pivoted.where('label = "sim100" and contigName = 1').head().values
+    assert len(pt_list) == 100
+    expected = [
```
Comment (karenfeng) [inline python/glow/wgr/tests/test_pivot_for_gwas.py:26]: Can you add a check that the list ordering matches the input sample ID order? This would also allow us to avoid hard-coding these values.

File: python/glow/wgr/functions.py [hunk 4]
```diff
@@ -128,3 +129,38 @@ def block_variants_and_samples(variant_df: DataFrame, sample_ids: List[str],
     output = blocked_gt, index_map
     assert check_return_type(output)
     return output
+
+
+def pivot_for_gwas(spark: SparkSession, label_df: pd.DataFrame) -> DataFrame:
+    """
+    Pivots a Pandas DataFrame into a Spark DataFrame with a convenient format for Glow's GWAS
```
Comment (karenfeng) [inline python/glow/wgr/functions.py:136]: I'm a little iffy about the term "pivot" here; generally, a pivot table contains summary statistics. I know Pandas uses `pivot` when reshaping though.

File: python/glow/wgr/tests/test_pivot_for_gwas.py (hunk 2 above)
Comment (henrydavidge) [reply python/glow/wgr/tests/test_pivot_for_gwas.py:20]: I actually don't think that we need to or should require this. If someone wants to treat contig names as ints, for instance, that should be totally fine.

**Comment by henrydavidge** [comment]
@karenfeng Comments addressed. Thanks for the testing suggestion -- that actually caught a bug 🤦 . Also, I changed the terminology to "reshape". Let me know if you like that better.

File: python/glow/wgr/functions.py [hunk 5]
```diff
@@ -157,7 +158,8 @@ def pivot_for_gwas(spark: SparkSession, label_df: pd.DataFrame) -> DataFrame:
         transposed_df = label_df.T
         column_names = ['label', 'values']
     elif label_df.index.nlevels == 2:  # Indexed by sample id and contig name
-        transposed_df = label_df.T.stack()
+        sample_ordering = pd.unique(label_df.index.get_level_values(0))
+        transposed_df = label_df.T.stack().reindex(sample_ordering, axis=1)
```
Comment (karenfeng) [inline python/glow/wgr/functions.py:162]: What happens if we use `.reindex(columns=labeldf.index)` here?

File: python/glow/wgr/functions.py [hunk 6]
```diff
@@ -157,7 +158,8 @@ def pivot_for_gwas(spark: SparkSession, label_df: pd.DataFrame) -> DataFrame:
         transposed_df = label_df.T
```
Comment (karenfeng) [inline python/glow/wgr/functions.py:158]: Does this guarantee ordering?

File: python/glow/wgr/functions.py (hunk 5 above)
Comment (henrydavidge) [reply python/glow/wgr/functions.py:162]: If the original index was a multi-index on sample id and contig name, so unfortunately that does not work. We need to the `unique` because the sample ids appear for each chromosome.

File: python/glow/wgr/functions.py (hunk 6 above)
Comment (henrydavidge) [reply python/glow/wgr/functions.py:158]: Yes, transposing does not sort the columns or rows.

File: python/glow/wgr/functions.py [hunk 7]
```diff
@@ -158,8 +158,10 @@ def reshape_for_gwas(spark: SparkSession, label_df: pd.DataFrame) -> DataFrame:
...
         column_names = ['label', 'values']
     elif label_df.index.nlevels == 2:  # Indexed by sample id and contig name
-        sample_ordering = pd.unique(label_df.index.get_level_values(0))
-        transposed_df = label_df.T.stack().reindex(sample_ordering, axis=1)
+        # stacking orders by the new index, so we need to remember the original sample order in case
```
Comment (karenfeng) [inline python/glow/wgr/functions.py:161]: I'm a little confused by the grammar here - can you rephrase?

Comment (henrydavidge) [reply python/glow/wgr/functions.py]: Done
